labgrid/driver/ubootdriver.py

```diff
@@ -69,26 +69,25 @@ def on_deactivate(self):
         """
         self._status = 0
 
-    def _run(self, cmd: str, *, timeout: int = 30, codec: str = "utf-8", decodeerrors: str = "strict"):  # pylint: disable=unused-argument,line-too-long
+    def _popen(self, cmd: str):
         # TODO: use codec, decodeerrors
         # TODO: Shell Escaping for the U-Boot Shell
         marker = gen_marker()
-        cmp_command = f"""echo '{marker[:4]}''{marker[4:]}'; {cmd}; echo "$?"; echo '{marker[:4]}''{marker[4:]}';"""  # pylint: disable=line-too-long
-        if self._status == 1:
-            self.console.sendline(cmp_command)
-            _, before, _, _ = self.console.expect(self.prompt, timeout=timeout)
-            # Remove VT100 Codes and split by newline
-            data = self.re_vt100.sub(
-                '', before.decode('utf-8'), count=1000000
-            ).replace("\r", "").split("\n")
-            self.logger.debug("Received Data: %s", data)
-            # Remove first element, the invoked cmd
-            data = data[data.index(marker) + 1:]
-            data = data[:data.index(marker)]
-            exitcode = int(data[-1])
-            del data[-1]
-            return (data, [], exitcode)
+        cmp_command = f"""echo '{marker[:4]}''{marker[4:]}'\\\\c; {cmd}; echo '{marker[:4]}''{marker[4:]}' $?;"""  # pylint: disable=line-too-long
+        self.console.sendline(cmp_command)
+        self.console.expect(marker)
+        return ConsoleMarkerProcess(self.console, marker, self.prompt)
 
+    def _run(self, cmd: str, *, timeout: int = 30, codec: str = "utf-8", decodeerrors: str = "strict"):  # pylint: disable=unused-argument,line-too-long
+        if self._status == 1:
+            with self._popen(cmd) as p:
+                output = p.read(timeout=timeout)
+                # Remove VT100 Codes and split by newline
+                data = self.re_vt100.sub(
+                    '', output.decode('utf-8'), count=1000000
+                ).replace("\r", "").split("\n")
+                self.logger.debug("Received Data: {data}")
```

Comment:
I guess this is meant as an f-string. Please convert to lazy % formatting.

Reply:
Per previous discussion, `%` is not lazt